### DISCUSSION TITLE
Extension making use of webview using the VSCode UI Toolkit - unable to manage a form and wait on completion

### DISCUSSION THREAD

**Discussion by davidkjackson54** [microsoft/vscode-discussions#647] [Extension Development QnA] [answered]
I have a VSCode Extension that has previously made use of the ShowInputBox feature but our users have complained as to its usage. when a number of fields are required to be provided.
So ,I have now implemented a webview  within my VSCode Extension and created a panel that requires a number of user input fields to be filled. However, there seems to be no obvious means of making the extension wait till the webview panel has been filled and the panel disposed of.   The extension should not be allowed to proceed till the field values are provided.  
With the existng ShowInputBox , it is easy to simply await on each request before proceeding.

I did open a bug to the VSCode folk but it was unceremoniously closed with 'no plan to implement". 

The inability to be able to manage a form containing some number of input fields is a real showstopper.

Now there may be some means of accomplishing this that I am entirely unaware of, but I have so far not found any workable solution.
Even being able to somehow track the on.dispose event  might be of some use but again the extension must be able to wait on that..

**Comment by r3m0t** [comment]
You can use `new Promise` to create promises that resolve on any particular event. Are you already sending messages from your webview with the field values? Is there a Submit button in the webview?

**Comment by davidkjackson54** [comment]
Yes, there is a submit button that validates the user supplied input, and I then sends message back from the webview with the completed input data as an object to my Extension code. 
What I haven't fathomed out is how to make the extension code wait till that occurs and I haven't found any such examples of how that might be accomplished.   
In addition to getting the data back from the webview, there is also an on.dispose event that triggers when the webview is destroyed but again I haven't been able to find any examples that might show how to wait on that.

**Comment by r3m0t** [comment]
This is probably simplest.

```typescript

// in your command:
const deferred = defer< {field1: string , field2: string}>( );
console.log("Showing form");
const myWebview = new MyWebview(param1, deferred );
const {field1, field2} = await deferred.promise;
console.log("User submitted", field1, field2);
// act on the user input

// in your MyWebview class:
webview.onMessage((message) => {
  deferred.resolve({field1: message.field1, field2: message.field2});
});
webview.onDispose(() => {
  deferred.reject(new Error("user cancelled"));
}
```

The `defer` function is available from https://www.npmjs.com/package/already

I would recommend following a tutorial on promises to understand `await` better and how you can create objects you can `await`.

**Reply by davidkjackson54** [reply]
Unfortunately this isn't proving to be that simple. I am familiar with async/await but this has me stumped.
 The extension invokes the command  via:
 ```
vscode.commands.executeCommand(
        "myExtension.createCredentials",
        profInfo,
        this.webviewBaseProfileArray,
        this.webviewSystemProfileArray
      );
``` 
That command is in extension.ts and it in turn fires off:
```
context.subscriptions.push(
    vscode.commands.registerCommand(
      "myExtensioonr.createCredentials",
      (profInfo, webviewbaseProfileArray, webviewzmfProfileArray) =>
        CreateCredentialsPanel.render(context.extensionUri, profInfo, webviewbaseProfileArray, webviewzmfProfileArray)
```
The render method in the CreateCredentials class invokes the constructor and it is the constructor that establishes the on.dispose event listener method and its own message listener for incoming message from the webview.  
The render method establishes the webview panel with createwebviewPanel() and posts the message (the data passed from the extension) to the webview code.
The webview panel (written in vue) picks up the message and does its thing. 
Finally, the user would click the button having entered the required input fields. The button onclick validates the data and then posts a message which is then picked up by the CreateCredentials messageListener and it finally calls the on dispose method to destroy the webview and terminates. 
I believe I need that at the time of the on.dispose being called that extension.ts which fired off the command in the first place to be sat waiting for a response and therefore having the extension sit and wait.    
I have spent a lot of time on this but I am now not seeing the wood for the trees.

**Reply by r3m0t** [reply, marked as answer]
This fits what I said:

```
const res = await vscode.commands.executeCommand(
       "myExtension.createCredentials",
       profInfo,
       this.webviewBaseProfileArray,
       this.webviewSystemProfileArray
     );
```

```context.subscriptions.push(
    vscode.commands.registerCommand(
      "myExtensioonr.createCredentials",
      (profInfo, webviewbaseProfileArray, webviewzmfProfileArray) =>
        CreateCredentialsPanel.render(context.extensionUri, profInfo, webviewbaseProfileArray, webviewzmfProfileArray)
```

Then in `render`, create the `deferred` variable and return `deferred.promise`. Pass `deferred` to the constructor you mentioned. You don't need to mark any functions `async` other than the first one.

By the way, within extensions you don't need to use `executeCommand/registerCommand`, you can just have direct function calls. That's a separate issue though.

**Reply by davidkjackson54** [reply]
@r3m0t  - Hi again, Thanks for the tip on the deferral. I came back to this this morning and have it working now.  
I wrote my own 'new Promise deferral' so didn't have to use the external package that you suggested and it is now working fine and awaits correctly.    Many thanks.

**Comment by chrisblanch39** [comment]
https://github.com/microsoft/vscode-discussions/discussions/647#discussioncomment-6204380
